File: backend/app/routers/inventory.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_user, get_db
from app.models.user import User
from app.schemas.inventory import (
    LeadInventoryRequirementCreate,
    LeadInventoryRequirementListResponse,
    LeadInventoryRequirementOut,
    LeadInventoryRequirementUpdate,
    InventoryTemplateApplyRequest,
    InventoryTemplateApplyResult,
    InventoryTemplateCreate,
    InventoryTemplateListResponse,
    InventoryTemplateOut,
    InventoryTemplateUpdate,
    InventoryItemCreate,
    InventoryItemListResponse,
    InventoryItemOut,
    InventoryItemUpdate,
    InventoryMovementCreate,
    InventoryMovementListResponse,
    InventoryMovementOut,
)
from app.services.business_service import BusinessService
from app.services.inventory_service import InventoryService
from app.utils.permissions import can_manage_inventory, can_view_inventory

router = APIRouter(prefix="/api/v1/inventory", tags=["inventory"])


def _item_out(item) -> InventoryItemOut:
    available_quantity = item.current_quantity - item.reserved_quantity
    return InventoryItemOut(
        id=item.id,
        business_id=item.business_id,
        name=item.name,
        sku=item.sku,
        unit=item.unit,
        current_quantity=item.current_quantity,
        reserved_quantity=item.reserved_quantity,
        available_quantity=available_quantity,
        min_quantity=item.min_quantity,
        notes=item.notes,
        is_active=item.is_active,
        created_at=item.created_at,
        low_stock=available_quantity <= item.min_quantity,
    )


async def _resolve_business_access(business_id: uuid.UUID, current_user: User, db: AsyncSession):
    business_service = BusinessService(db)
    business = await business_service.get_for_user(business_id, current_user.id)
    if business is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Business not found")
    if "inventory" not in (business.enabled_modules or []):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Inventory module is not enabled for this business")
    member = await business_service.get_member(business_id, current_user.id)
    role = member.role if member else None
    custom_permissions = member.custom_permissions if member else []
    return business, role, custom_permissions


@router.get("/items", response_model=InventoryItemListResponse)
async def list_inventory_items(
    business_id: uuid.UUID = Query(...),
    include_inactive: bool = Query(default=False),
    search: str = Query(default=""),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryItemListResponse:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_view_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot view inventory.")
    items = await InventoryService(db).list_items(business_id, include_inactive=include_inactive, search=search.strip())
    return InventoryItemListResponse(items=[_item_out(item) for item in items])


@router.post("/items", response_model=InventoryItemOut)
async def create_inventory_item(
    payload: InventoryItemCreate,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryItemOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot create inventory items.")
    item = await InventoryService(db).create_item(business_id, payload, triggered_by_user_id=current_user.id)
    return _item_out(item)


@router.patch("/items/{item_id}", response_model=InventoryItemOut)
async def update_inventory_item(
    item_id: uuid.UUID,
    payload: InventoryItemUpdate,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryItemOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot update inventory items.")
    service = InventoryService(db)
    item = await service.get_item(business_id, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory item not found")
    updated = await service.update_item(item, payload, triggered_by_user_id=current_user.id)
    return _item_out(updated)


@router.get("/movements", response_model=InventoryMovementListResponse)
async def list_inventory_movements(
    business_id: uuid.UUID = Query(...),
    item_id: uuid.UUID | None = Query(default=None),
    lead_id: uuid.UUID | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryMovementListResponse:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_view_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot view inventory movements.")
    items = await InventoryService(db).list_movements(business_id, item_id=item_id, lead_id=lead_id)
    return InventoryMovementListResponse(items=[InventoryMovementOut.model_validate(item) for item in items])


@router.post("/items/{item_id}/movements", response_model=InventoryMovementOut)
async def create_inventory_movement(
    item_id: uuid.UUID,
    payload: InventoryMovementCreate,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryMovementOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory movements.")
    service = InventoryService(db)
    item = await service.get_item(business_id, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory item not found")
    try:
        movement = await service.add_movement(business_id, item, current_user.id, payload)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return InventoryMovementOut.model_validate(movement)


@router.get("/requirements", response_model=LeadInventoryRequirementListResponse)
async def list_lead_inventory_requirements(
    business_id: uuid.UUID = Query(...),
    lead_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> LeadInventoryRequirementListResponse:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_view_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot view inventory requirements.")
    items = await InventoryService(db).list_requirements(business_id, lead_id)
    return LeadInventoryRequirementListResponse(items=[LeadInventoryRequirementOut.model_validate(item) for item in items])


@router.post("/requirements", response_model=LeadInventoryRequirementOut)
async def create_lead_inventory_requirement(
    payload: LeadInventoryRequirementCreate,
    business_id: uuid.UUID = Query(...),
    lead_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> LeadInventoryRequirementOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory requirements.")
    item = await InventoryService(db).create_requirement(business_id, lead_id, payload)
    return LeadInventoryRequirementOut.model_validate(item)


@router.patch("/requirements/{requirement_id}", response_model=LeadInventoryRequirementOut)
async def update_lead_inventory_requirement(
    requirement_id: uuid.UUID,
    payload: LeadInventoryRequirementUpdate,
    business_id: uuid.UUID = Query(...),
    lead_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> LeadInventoryRequirementOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory requirements.")
    service = InventoryService(db)
    requirement = await service.get_requirement(business_id, lead_id, requirement_id)
    if requirement is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory requirement not found")
    updated = await service.update_requirement(requirement, payload)
    return LeadInventoryRequirementOut.model_validate(updated)


@router.delete("/requirements/{requirement_id}")
async def delete_lead_inventory_requirement(
    requirement_id: uuid.UUID,
    business_id: uuid.UUID = Query(...),
    lead_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, str]:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory requirements.")
    service = InventoryService(db)
    requirement = await service.get_requirement(business_id, lead_id, requirement_id)
    if requirement is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory requirement not found")
    await service.delete_requirement(requirement)
    return {"message": "Inventory requirement removed."}


@router.get("/templates", response_model=InventoryTemplateListResponse)
async def list_inventory_templates(
    business_id: uuid.UUID = Query(...),
    event_type_match: str | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryTemplateListResponse:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_view_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot view inventory templates.")
    items = await InventoryService(db).list_templates(business_id, event_type_match=event_type_match.strip() if event_type_match else None)
    return InventoryTemplateListResponse(items=[InventoryTemplateOut.model_validate(item) for item in items])


@router.post("/templates", response_model=InventoryTemplateOut)
async def create_inventory_template(
    payload: InventoryTemplateCreate,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryTemplateOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory templates.")
    item = await InventoryService(db).create_template(business_id, payload)
    return InventoryTemplateOut.model_validate(item)


@router.delete("/templates/{template_id}")
async def delete_inventory_template(
    template_id: uuid.UUID,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, str]:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory templates.")
    service = InventoryService(db)
    template = await service.get_template(business_id, template_id)
    if template is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory template not found")
    await service.delete_template(template)
    return {"message": "Inventory template removed."}


@router.patch("/templates/{template_id}", response_model=InventoryTemplateOut)
async def update_inventory_template(
    template_id: uuid.UUID,
    payload: InventoryTemplateUpdate,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryTemplateOut:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot manage inventory templates.")
    service = InventoryService(db)
    template = await service.get_template(business_id, template_id)
    if template is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory template not found")
    updated = await service.update_template(template, payload)
    return InventoryTemplateOut.model_validate(updated)


@router.post("/templates/{template_id}/apply", response_model=InventoryTemplateApplyResult)
async def apply_inventory_template(
    template_id: uuid.UUID,
    payload: InventoryTemplateApplyRequest,
    business_id: uuid.UUID = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> InventoryTemplateApplyResult:
    _business, role, custom_permissions = await _resolve_business_access(business_id, current_user, db)
    if not can_manage_inventory(role, custom_permissions):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You cannot apply inventory templates.")
    service = InventoryService(db)
    template = await service.get_template(business_id, template_id)
    if template is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Inventory template not found")
    try:
        result = await service.apply_template(
            business_id,
            payload.lead_id,
            template,
            current_user.id,
        )
    except ValueError as exc:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return InventoryTemplateApplyResult(**result)

File: miniapp/src/api/inventory.ts
import { apiClient } from "./client";

export type InventoryItem = {
  id: string;
  business_id: string;
  name: string;
  sku?: string | null;
  unit: string;
  current_quantity: string;
  reserved_quantity: string;
  available_quantity: string;
  min_quantity: string;
  notes?: string | null;
  is_active: boolean;
  created_at: string;
  low_stock: boolean;
};

export type InventoryMovement = {
  id: string;
  business_id: string;
  item_id: string;
  lead_id?: string | null;
  created_by: string;
  movement_type: string;
  quantity: string;
  note?: string | null;
  created_at: string;
};

export type LeadInventoryRequirement = {
  id: string;
  business_id: string;
  lead_id: string;
  item_id: string;
  required_quantity: string;
  note?: string | null;
  created_at: string;
};

export type InventoryTemplate = {
  id: string;
  business_id: string;
  name: string;
  event_type_match?: string | null;
  note?: string | null;
  items: Array<{
    item_id: string;
    required_quantity: string;
    note?: string | null;
  }>;
  created_at: string;
};

export type InventoryTemplateApplyResult = {
  requirements_created: number;
  reserved_units: string;
  missing_units: string;
  prep_task_created: boolean;
  restock_task_created: boolean;
  message: string;
};

export async function getInventoryItems(
  businessId: string,
  token?: string | null,
  search = "",
) {
  const response = await apiClient.get<{ items: InventoryItem[] }>("/api/v1/inventory/items", {
    params: {
      business_id: businessId,
      search: search || undefined,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function createInventoryItem(
  businessId: string,
  payload: {
    name: string;
    sku?: string | null;
    unit?: string;
    current_quantity?: string;
    min_quantity?: string;
    notes?: string | null;
  },
  token?: string | null,
) {
  const response = await apiClient.post<InventoryItem>("/api/v1/inventory/items", payload, {
    params: { business_id: businessId },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function updateInventoryItem(
  businessId: string,
  itemId: string,
  payload: {
    name?: string;
    sku?: string | null;
    unit?: string;
    min_quantity?: string;
    notes?: string | null;
    is_active?: boolean;
  },
  token?: string | null,
) {
  const response = await apiClient.patch<InventoryItem>(`/api/v1/inventory/items/${itemId}`, payload, {
    params: { business_id: businessId },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function getInventoryMovements(
  businessId: string,
  token?: string | null,
  itemId?: string | null,
  leadId?: string | null,
) {
  const response = await apiClient.get<{ items: InventoryMovement[] }>("/api/v1/inventory/movements", {
    params: {
      business_id: businessId,
      item_id: itemId || undefined,
      lead_id: leadId || undefined,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function getInventoryMovementsForLead(
  businessId: string,
  token?: string | null,
  leadId?: string | null,
) {
  return getInventoryMovements(businessId, token, null, leadId);
}

export async function createInventoryMovement(
  businessId: string,
  itemId: string,
  payload: {
    movement_type: string;
    quantity: string;
    lead_id?: string | null;
    note?: string | null;
  },
  token?: string | null,
) {
  const response = await apiClient.post<InventoryMovement>(`/api/v1/inventory/items/${itemId}/movements`, payload, {
    params: { business_id: businessId },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function getLeadInventoryRequirements(
  businessId: string,
  leadId: string,
  token?: string | null,
) {
  const response = await apiClient.get<{ items: LeadInventoryRequirement[] }>("/api/v1/inventory/requirements", {
    params: {
      business_id: businessId,
      lead_id: leadId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function createLeadInventoryRequirement(
  businessId: string,
  leadId: string,
  payload: {
    item_id: string;
    required_quantity: string;
    note?: string | null;
  },
  token?: string | null,
) {
  const response = await apiClient.post<LeadInventoryRequirement>("/api/v1/inventory/requirements", payload, {
    params: {
      business_id: businessId,
      lead_id: leadId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function deleteLeadInventoryRequirement(
  businessId: string,
  leadId: string,
  requirementId: string,
  token?: string | null,
) {
  const response = await apiClient.delete<{ message: string }>(`/api/v1/inventory/requirements/${requirementId}`, {
    params: {
      business_id: businessId,
      lead_id: leadId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function getInventoryTemplates(
  businessId: string,
  token?: string | null,
  eventTypeMatch?: string | null,
) {
  const response = await apiClient.get<{ items: InventoryTemplate[] }>("/api/v1/inventory/templates", {
    params: {
      business_id: businessId,
      event_type_match: eventTypeMatch || undefined,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function createInventoryTemplate(
  businessId: string,
  payload: {
    name: string;
    event_type_match?: string | null;
    note?: string | null;
    items: Array<{
      item_id: string;
      required_quantity: string;
      note?: string | null;
    }>;
  },
  token?: string | null,
) {
  const response = await apiClient.post<InventoryTemplate>("/api/v1/inventory/templates", payload, {
    params: {
      business_id: businessId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function updateInventoryTemplate(
  businessId: string,
  templateId: string,
  payload: {
    name?: string;
    event_type_match?: string | null;
    note?: string | null;
    items?: Array<{
      item_id: string;
      required_quantity: string;
      note?: string | null;
    }>;
  },
  token?: string | null,
) {
  const response = await apiClient.patch<InventoryTemplate>(`/api/v1/inventory/templates/${templateId}`, payload, {
    params: {
      business_id: businessId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function deleteInventoryTemplate(
  businessId: string,
  templateId: string,
  token?: string | null,
) {
  const response = await apiClient.delete<{ message: string }>(`/api/v1/inventory/templates/${templateId}`, {
    params: {
      business_id: businessId,
    },
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });
  return response.data;
}

export async function applyInventoryTemplate(
  businessId: string,
  templateId: string,
  leadId: string,
  token?: string | null,
) {
  const response = await apiClient.post<InventoryTemplateApplyResult>(
    `/api/v1/inventory/templates/${templateId}/apply`,
    { lead_id: leadId },
    {
      params: {
        business_id: businessId,
      },
      headers: token ? { Authorization: `Bearer ${token}` } : undefined,
    },
  );
  return response.data;
}
